README.md

```diff
@@ -647,14 +647,382 @@ type VirtualMachine struct {
+            pcUpdate = PcUpdateJumpRel
+        case 4:
+            pcUpdate = PcUpdateJnz
+        default:
+            return Instruction{}, ErrInvalidPcUpdateError
+	}
+
+	switch resLogicNum {
+        case 0:
+            if pcUpdate == PcUpdateJnz {
+                res = ResUnconstrained
+            } else {
+                res = ResOp1
+            }
+        case 1:
+            res = ResAdd
+        case 2:
+            res = ResMul
+        default:
+            return Instruction{}, ErrInvalidResError
+	}
+
+	switch opCodeNum {
+        case 0:
+            opcode = NOp
+        case 1:
+            opcode = Call
+        case 2:
+            opcode = Ret
+        case 4:
+            opcode = AssertEq
+        default:
+            return Instruction{}, ErrInvalidOpcodeError
+	}
+
+	switch apUpdateNum {
+        case 0:
+            if opcode == Call {
+                apUpdate = ApUpdateAdd2
+            } else {
+                apUpdate = ApUpdateRegular
+            }
+        case 1:
+            apUpdate = ApUpdateAdd
+        case 2:
+            apUpdate = ApUpdateAdd1
+        default:
+            return Instruction{}, ErrInvalidApUpdateError
+	}
+
+	switch opcode {
+        case Call:
+            fpUpdate = FpUpdateAPPlus2
+        case Ret:
+            fpUpdate = FpUpdateDst
+        default:
+            fpUpdate = FpUpdateRegular
+	}
+    ```
+    
+6. Creating the Instruction and Returning the Result:
+    With all the necessary information extracted and decoded, the method constructs an Instruction object by assigning the decoded values to its fields and returns the created Instruction object along with a nil error if the decoding process is successful.
+
+    ```go 
+    return Instruction{
+		Off0:     offset0,
+		Off1:     offset1,
+		Off2:     offset2,
+		DstReg:   dstRegister,
+		Op0Reg:   op0Register,
+		Op1Addr:  op1Src,
+		ResLogic: res,
+		PcUpdate: pcUpdate,
+		ApUpdate: apUpdate,
+		FpUpdate: fpUpdate,
+		Opcode:   opcode,
+	}, nil
+    ```
+7. Error Handling:
+    If at any point during the decoding process, an unexpected value is encountered or the input doesn't conform to the expected pattern, the method returns an appropriate error. These errors include cases like invalid op1 register, invalid pc update, invalid result logic, invalid opcode, etc.
+
+    ```go
+    var ErrNonZeroHighBitError = errors.New("Instruction high bit was not set to zero")
+    var ErrInvalidOp1RegError = errors.New("Instruction had invalid Op1 Register")
+    var ErrInvalidPcUpdateError = errors.New("Instruction had invalid Pc update")
+    var ErrInvalidResError = errors.New("Instruction had an invalid res")
+    var ErrInvalidOpcodeError = errors.New("Instruction had an invalid opcode")
+    var ErrInvalidApUpdateError = errors.New("Instruction had an invalid Ap Update")
+    ```
 #### Run instruction
 
+At this point, we have stored all the needed information of the instruction in our program's. Let's run it! 
```

Comment:
```suggestion
At this point, we have stored all the needed information of the instruction in our program's. Let's run it! 
```
This line is confusing, are we storing something in the program?
Maybe we could simplify it to "At this point, we have all the information we need from the instruction. Let's run it!"